### PR TITLE
Desul: Add ScopeCaller

### PR DESCRIPTION
**Pull request by crtrott**
In Serial only builds Kokkos::atomics will use ScopeCaller

### CODE REVIEW

**Comment by crtrott** [comment]
Corresponding PR in Desul: https://github.com/desul/desul/pull/50

Note I applied desul clang-format too

File: core/src/desul/atomics/Compare_Exchange_GCC.hpp [hunk 1]
```diff
@@ -18,72 +18,71 @@ SPDX-License-Identifier: (BSD-3-Clause)
...
-} //namespace Impl
+}  // namespace Impl
 
-#if defined(__clang__) && (__clang_major__>=7) && !defined(__APPLE__)
+#if defined(__clang__) && (__clang_major__ >= 7) && !defined(__APPLE__)
 // Disable warning for large atomics on clang 7 and up (checked with godbolt)
-// error: large atomic operation may incur significant performance penalty [-Werror,-Watomic-alignment]
-// https://godbolt.org/z/G7YhqhbG6
+// error: large atomic operation may incur significant performance penalty
+// [-Werror,-Watomic-alignment] https://godbolt.org/z/G7YhqhbG6
```
Comment (dalg24) [inline core/src/desul/atomics/Compare_Exchange_GCC.hpp:38]:
```suggestion
// clang-format off
// error: large atomic operation may incur significant performance penalty [-Werror,-Watomic-alignment]
// clang-format on
// https://godbolt.org/z/G7YhqhbG6
```
Uggg.  Did we merge that into Desul?

**Review by dalg24 — COMMENTED** [review]
Would you do the formatting part in a separate PR?

File: core/src/desul/atomics/Compare_Exchange_GCC.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/desul/atomics/Compare_Exchange_GCC.hpp]: yeah

File: core/src/desul/atomics/Compare_Exchange_ScopeCaller.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    T current_val = *dest;                                            \
+    if (current_val == compare) *dest = value;                        \
+    return current_val;                                               \
+  }
+
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderSeqCst)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderAcqRel)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderRelease)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderAcquire)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderRelaxed)
```
Comment (dalg24) [inline core/src/desul/atomics/Compare_Exchange_ScopeCaller.hpp:39]: Why can't this be templated on the memory order?

**Review by dalg24 — COMMENTED** [review]
I'd prefer if you submitted the changes to `<desul/atomics/Generic.hpp>` in another PR

**Review by masterleinad — APPROVED** [review]
Looks OK to me (despite that the name feels unintuitive at least for the current use case to me). We at least have the Windows build and the OSX builds to test this.

File: core/src/desul/atomics/Compare_Exchange_ScopeCaller.hpp (hunk 2 above)
Comment (crtrott) [reply core/src/desul/atomics/Compare_Exchange_ScopeCaller.hpp:39]:
```
C:\projects\source\core\src\Kokkos_Atomics_Desul_Wrapper.hpp(226,17): error C2668: 'desul::atomic_compare_exchange': ambiguous call to overloaded function [C:\projects\source\build\core\src\kokkoscore.vcxproj]
C:\projects\source\core\src\desul/atomics/Compare_Exchange_MSVC.hpp(144,50): message : could be 'int desul::atomic_compare_exchange<T,desul::MemoryScopeCaller>(T *const ,T,T,desul::MemoryOrderRelaxed,MemoryScope)' [C:\projects\source\build\core\src\kokkoscore.vcxproj]
          with
          [
              T=int,
              MemoryScope=desul::MemoryScopeCaller
          ]
C:\projects\source\core\src\desul/atomics/Compare_Exchange_ScopeCaller.hpp(28,25): message : or       'T desul::atomic_compare_exchange<T,desul::MemoryOrderRelaxed>(T *,T,T,MemoryOrder,desul::MemoryScopeCaller)' [C:\projects\source\build\core\src\kokkoscore.vcxproj]
```

Comment (crtrott) [reply core/src/desul/atomics/Compare_Exchange_ScopeCaller.hpp:39]: The above error is for the Serial only build with MSVC and comes from the fact that the MSVC compare exchange is templated on Scope, but explicitly instantiated for various memory orders and thus both variants have one template parameter.

**Comment by dalg24** [comment]
I'll cut you a deal.  I close my eyes and merge this PR as is and you agree to split up your PR on the Desul side.
What do you say?

**Comment by crtrott** [comment]
Sure Damien :-)

**Comment by ajpowelsnl** [comment]
Morning @dalg24 and @crtrott -- is this issue still a blocker (given that it's been merged)?

**Comment by dalg24** [comment]
This is a pull request not an issue and it has been merged.

**Pull request merged**